src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var targetNodes = diagnostics.SelectAsArray(
+                d => root.FindNode(d.Location.SourceSpan, getInnermostNodeForTie: true)
+                         .GetAncestorsOrThis<ExpressionSyntax>().FirstOrDefault());
+
+            await editor.ApplyExpressionLevelSemanticEditsAsync(
+                document, targetNodes,
+                (semanticModel, targetNode) => true,
+                (semanticModel, currentRoot, targetNode) =>
+                {
+                    // All diagnostics have the same error code
+                    if (TryGetTargetTypeInfo(semanticModel, diagnostics[0].Id, targetNode, cancellationToken, out var nodeType, out var potentialConversionTypes)
+                        && potentialConversionTypes.Length == 1)
+                    {
+                        return ApplyFix(currentRoot, targetNode, potentialConversionTypes[0]);
+                    }
+
+                    return currentRoot;
+                },
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument, equivalenceKey: title)
+            {
+            }
+        }
+
+        private sealed class InheritanceDistanceComparer : IComparer<ITypeSymbol>
+        {
+            private readonly ITypeSymbol _baseType;
+            private readonly SemanticModel _semanticModel;
+
+            private int GetInheritanceDistance(ITypeSymbol baseType, ITypeSymbol? derivedType)
+            {
+                if (derivedType == null) return int.MaxValue;
+                if (derivedType.Equals(baseType)) return 0;
+
+                var distance = GetInheritanceDistance(baseType, derivedType.BaseType);
+
+                if (derivedType.Interfaces.Length != 0)
+                {
+                    foreach (var interfaceType in derivedType.Interfaces)
+                    {
+                        distance = Math.Min(GetInheritanceDistance(baseType, interfaceType), distance);
+                    }
+                }
+
+                return distance == int.MaxValue ? distance : distance + 1;
+            }
+            public int Compare(ITypeSymbol x, ITypeSymbol y)
+            {
+                // if the node has the explicit conversion operator, then it has the shortest distance
+                var xComversion = _semanticModel.Compilation.ClassifyCommonConversion(_baseType, x);
+                var xDist = xComversion.IsUserDefined || xComversion.IsNumeric ?
+                    0 : GetInheritanceDistance(_baseType, x);
```

Comment:
why are user-defined and numeric conversions a distance of 0?  that needs to be explained.